crates/bevy_ptr/src/lib.rs

```diff
@@ -11,26 +11,142 @@ use core::{
     cell::UnsafeCell,
     fmt::{self, Debug, Formatter, Pointer},
     marker::PhantomData,
-    mem::ManuallyDrop,
+    mem::{self, ManuallyDrop, MaybeUninit},
     num::NonZeroUsize,
     ptr::{self, NonNull},
 };
 
-/// Used as a type argument to [`Ptr`], [`PtrMut`] and [`OwningPtr`] to specify that the pointer is aligned.
+/// Used as a type argument to [`Ptr`], [`PtrMut`], [`OwningPtr`], and [`MovingPtr`] to specify that the pointer is guaranteed
+/// to be [aligned].
+///
+/// [aligned]: https://doc.rust-lang.org/std/ptr/index.html#alignment
 #[derive(Debug, Copy, Clone)]
 pub struct Aligned;
 
-/// Used as a type argument to [`Ptr`], [`PtrMut`] and [`OwningPtr`] to specify that the pointer is not aligned.
+/// Used as a type argument to [`Ptr`], [`PtrMut`], [`OwningPtr`], and [`MovingPtr`] to specify that the pointer may not [aligned].
+///
+/// [aligned]: https://doc.rust-lang.org/std/ptr/index.html#alignment
 #[derive(Debug, Copy, Clone)]
 pub struct Unaligned;
 
 /// Trait that is only implemented for [`Aligned`] and [`Unaligned`] to work around the lack of ability
 /// to have const generics of an enum.
-pub trait IsAligned: sealed::Sealed {}
+pub trait IsAligned: sealed::Sealed {
+    /// Reads the value pointed to by `ptr`.
+    ///
+    /// # Safety
+    ///  - `ptr` must be valid for reads.
+    ///  - `ptr` must point to a valid instance of type `T`
+    ///  - If this type is [`Aligned`], then `ptr` must be [properly aligned] for type `T`.
+    ///
+    /// [properly aligned]: https://doc.rust-lang.org/std/ptr/index.html#alignment
+    #[doc(hidden)]
+    unsafe fn read_ptr<T>(ptr: *const T) -> T;
+
+    /// Copies `count * size_of::<T>()` bytes from `src` to `dst`. The source
+    /// and destination must *not* overlap.
+    ///
+    /// # Safety
+    ///  - `src` must be valid for reads of `count * size_of::<T>()` bytes.
+    ///  - `dst` must be valid for writes of `count * size_of::<T>()` bytes.
+    ///  - The region of memory beginning at `src` with a size of `count *
+    ///    size_of::<T>()` bytes must *not* overlap with the region of memory
+    ///    beginning at `dst` with the same size.
+    ///  - If this type is [`Aligned`], then both `src` and `dst` must properly
+    ///    be aligned for values of type `T`.
+    #[doc(hidden)]
+    unsafe fn copy_nonoverlapping<T>(src: *const T, dst: *mut T, size: usize);
```

Comment:
Here the function argument is called `size` but the docs talk about `count` and it's not clear to me if it is supposed to be `size == count` or `size == count * size_of::<T>()`.